### PR TITLE
Heading and onSubmit a11y adjustments

### PR DESCRIPTION
**Pull request by amy-corson-ibigroup**
- Adds `tabIndex` and `aria-assertive` to headings to ensure they are announced to users relying on AT.
- Changes question titles to `h1` elements
- "Next" button changes to "Submit" and has `type="submit"` on survey last question.
- Navigation buttons moved into separate component to limit `index.tsx` complexity.
- `h1` elements moved out of question components
- Focus set on `h1` when question is changed

### CODE REVIEW

File: components/NavButtons.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React from 'react'
+import { useTranslations } from 'next-intl'
+
+type Props = {
+  activeQuestion: number
+  connected: boolean
+  lastQuestion: boolean
```
Comment (miles-grant-ibigroup) [inline components/NavButtons.tsx]: We should probably rename this to something like `isLastQuestion` so it's not confused for an action

File: components/NavButtons.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  surveyOver: boolean
+}
+
+const NavButtons = ({
+  activeQuestion,
+  connected,
+  lastQuestion,
+  nextQuestion,
+  prevQuestion,
+  surveyOver
```
Comment (miles-grant-ibigroup) [inline components/NavButtons.tsx]: `activeQuestion`, `connected`, and `surveyOver` are only used to disable the buttons so it might make more sense to make a new prop `nextDisabled` `prevDisabled` and then do these checks outside the component? Or you could get really fancy and say if `nextQuestion` is null the button gets disabled but I'd document that carefully

**Review by miles-grant-ibigroup — APPROVED** [review]
This looks fantastic! Great stuff. Just a few nitpicks but nothing serious

File: components/NavButtons.tsx (hunk 2 above)
Comment (amy-corson-ibigroup) [reply components/NavButtons.tsx]: Ooh good idea!!

**Comment by amy-corson-ibigroup** [comment]
Assigning this back to me - the `h1` elements should not only be focusable, the focus should be set to them when each new question renders!

File: components/QuestionRenderer.tsx [hunk 3]
```diff
@@ -41,78 +47,107 @@ const QuestionRenderer = ({
...
+            {title}
+          </h1>
+        )}
+        {subtitle && <h2>{subtitle}</h2>}
+      </>
+    )
+  }
+
+  return (
+    <>
```
Comment (miles-grant-ibigroup) [inline components/QuestionRenderer.tsx:130]: You can get rid of this, the fieldset is the outermost element

File: components/SatisfactionSlider.tsx [hunk 4]
```diff
@@ -41,35 +41,27 @@ const SatisfactionSlider = ({
...
-          name={title}
-          onChange={(e) => updateNumber(parseInt(e.target.value))}
-          step={step}
-          type="range"
-          value={number}
-        />
-      </fieldset>
+      <span className={styles.invisibleA11yLabel}>
+        {t('Slider.valueDescription')}
+      </span>{' '}
```
Comment (miles-grant-ibigroup) [inline components/SatisfactionSlider.tsx]: Do we need this space?

File: components/SatisfactionSlider.tsx [hunk 5]
```diff
@@ -41,35 +41,27 @@ const SatisfactionSlider = ({
...
+      <Smiley percentage={scaledNumber} />
+      <input
+        aria-hidden="true"
+        disabled={disabled}
+        name={title}
+        type="hidden"
+        value={scaledNumber}
+      />
```
Comment (miles-grant-ibigroup) [inline components/SatisfactionSlider.tsx]: Why is this here? If it is hidden and disabled

File: components/Stars.tsx [hunk 6]
```diff
@@ -5,18 +5,18 @@ import starStyles from '../styles/Stars.module.css'
...
-const Star = ({ selected }: { selected?: boolean }) => {
+const Star = ({ index, selected }: { index?: string; selected?: boolean }) => {
   let fill = 'rgba(30, 30, 30, 0.3)'
   if (selected) {
     fill = 'rgba(255, 214, 0, 0.8)'
   }
 
   return (
     <svg
+      aria-label={index ? `${index + 1} stars` : ''}
```
Comment (miles-grant-ibigroup) [inline components/Stars.tsx:19]: If you take a look at like 64 of this file you'll see how to get the current star index without needing to pass it in explicitly

File: pages/index.tsx [hunk 7]
```diff
@@ -48,16 +49,24 @@ const Home: NextPage = ({ questions, socketServerUrl }: Props) => {
     })
   }
 
+  // On each question render, set focus to h1 for AT
+  useEffect(() => {
+    headingRefs.current[activeQuestion]?.focus()
```
Comment (miles-grant-ibigroup) [inline pages/index.tsx]: Not sure if you tried this already, but what do you think about creating a new `autoFocusH1` component which focuses itself on render? That way we don't need all this ref infrastructure

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
A lot of scary changes here, but most of the changes make sense!

I've done a first pass on some details which might help make the changes more manageable

File: pages/index.tsx (hunk 7 above)
Comment (amy-corson-ibigroup) [reply pages/index.tsx]: I think I tried what you're talking about, but the tricky thing is all of these headers get rendered at the same time, so there has to be some way to pass the focus based on the active question! Do you think there's a way to use an autofocus component without the multiple refs?

Comment (miles-grant-ibigroup) [reply pages/index.tsx]: This is a classic react problem. It's a bit of a hack but what do you think about a handler on the current elements that automatically focuses the parent element when it's selected?

Also a bit annoyed at this change -- the textbox auto focusing was a usability feature. Any way we can make that configurable?

File: components/SatisfactionSlider.tsx (hunk 5 above)
Comment (amy-corson-ibigroup) [reply components/SatisfactionSlider.tsx]: Good catch, this was moving the `disabled` from the `fieldset` to the `input`! Just moved it to the wrong one whoops.

File: pages/index.tsx (hunk 7 above)
Comment (amy-corson-ibigroup) [reply pages/index.tsx]: Per our offline conversation, I went ahead and removed setting the focus to the `h1` with the thought being that the `aria-live` & the focusable `h1` will be sufficient

File: components/Stars.tsx (hunk 6 above)
Comment (amy-corson-ibigroup) [reply components/Stars.tsx:19]: I tried both this approach and using a ref to refer to the id, but since the id of each star element is not simply the index number, it wound up adding a lot of code! I THINK passing the index might be more efficient but happy to further talk about this!

**Review by miles-grant-ibigroup — APPROVED** [review]
Sad that we lose the cool `title` prop but all works for me and aria viewer is happy!

File: components/NavButtons.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    <>
+      <button disabled={prevDisabled} onClick={prevQuestion}>
+        {t('index.prev')}
+      </button>
+      <button
+        disabled={nextDisabled}
+        onClick={nextQuestion}
+        type={isLastQuestion ? 'submit' : 'button'}
+      >
+        {isLastQuestion ? t('index.submitFeedback') : t('index.next')}
```
Comment (binh-dam-ibigroup) [inline components/NavButtons.tsx:30]: `'index.submitFeedback'` is missing English content in the test deployments.

File: components/SatisfactionSlider.tsx [hunk 9]
```diff
@@ -41,35 +41,27 @@ const SatisfactionSlider = ({
...
-          min={min}
-          name={title}
-          onChange={(e) => updateNumber(parseInt(e.target.value))}
-          step={step}
-          type="range"
-          value={number}
-        />
-      </fieldset>
+      <span className={styles.invisibleA11yLabel}>
+        {t('Slider.valueDescription')}
```
Comment (binh-dam-ibigroup) [inline components/SatisfactionSlider.tsx]: `'Slider.valueDescription'` is missing English content in the test deployments.

File: components/QuestionRenderer.tsx [hunk 10]
```diff
@@ -41,78 +44,101 @@ const QuestionRenderer = ({
...
+      </>
+    )
+  }
+
+  return (
+    <fieldset className="container">
+      {title && (
+        <legend>
+          <h1
+            aria-live="assertive"
```
Comment (binh-dam-ibigroup) [inline components/QuestionRenderer.tsx]: Is this causing VoiceOver on Safari to repeat the question at least twice?

File: components/SatisfactionSlider.tsx [hunk 11]
```diff
@@ -41,35 +41,27 @@ const SatisfactionSlider = ({
...
+      </span>
+      <Smiley percentage={scaledNumber} />
+      <input
+        aria-hidden="true"
```
Comment (binh-dam-ibigroup) [inline components/SatisfactionSlider.tsx]: Is `aria-hidden` needed for `<input type="hidden">`?

File: components/SatisfactionSlider.tsx [hunk 12]
```diff
@@ -41,35 +41,27 @@ const SatisfactionSlider = ({
...
+        type="hidden"
+        value={scaledNumber}
+      />
+      <input
+        aria-label={title}
+        disabled={disabled}
+        max={max}
+        min={min}
+        name={title}
+        onChange={(e) => updateNumber(parseInt(e.target.value))}
```
Comment (binh-dam-ibigroup) [inline components/SatisfactionSlider.tsx]: Consider `useCallback` (multiple instances in this PR).

File: components/Stars.tsx [hunk 13]
```diff
@@ -50,40 +49,38 @@ const Stars = ({
...
-          ))}
-      </fieldset>
-    </>
+    <div className={starStyles.container}>
+      {Array(number)
+        .fill(null)
+        .map((_, index) => (
+          <React.Fragment key={index}>
+            <input
+              className={starStyles.hidden}
```
Comment (binh-dam-ibigroup) [inline components/Stars.tsx]: This makes it impossible to select stars using the keyboard.

File: components/TextResponse.tsx [hunk 14]
```diff
@@ -30,28 +30,21 @@ const TextResponse = ({
...
-        />
-      </fieldset>
+      <textarea
+        className={`${styles.textbox} ${disabled ? 'disabled' : ''}`}
+        disabled={disabled}
+        id={title}
+        name={title}
+        onChange={(e) => updateUserText(e.target.value)}
+        placeholder={placeholder}
+        // Focus the textarea on render
```
Comment (binh-dam-ibigroup) [inline components/TextResponse.tsx:42]: Can controls on every question type be focused on render too?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Some content is missing, and could you check keyboard a11y and why questions are repeated multiple times with VoiceOver?

File: components/QuestionRenderer.tsx (hunk 10 above)
Comment (amy-corson-ibigroup) [reply components/QuestionRenderer.tsx]: I think the mac screenreader reads legends, but perhaps JAWS or other AT doesn't?

File: components/NavButtons.tsx (hunk 8 above)
Comment (amy-corson-ibigroup) [reply components/NavButtons.tsx:30]: I think that will require an update to the ST and ATL survey configs, which will be a separate PR, but correct me if I'm wrong!

File: components/SatisfactionSlider.tsx (hunk 9 above)
Comment (amy-corson-ibigroup) [reply components/SatisfactionSlider.tsx]: Same here, I think this should be fixed in the survey configs.

File: components/QuestionRenderer.tsx (hunk 10 above)
Comment (amy-corson-ibigroup) [reply components/QuestionRenderer.tsx]: Per accessibility consultant, "This announcement bug is a known issue which hasn’t been addressed by Apple (more info here: https://github.com/alphagov/govuk-frontend/issues/2395."

The suggestion was to move the `h1` heading outside the `<legend>` element. "Even though the text is repeated in both elements, that will at least ensure that each respective element is announced more contextually (e.g. the `h1` being announced as a heading, and the `<legend>` text being announced in context with the fields to which it relates.)"

Feedback applied in [0adff38](https://github.com/ibi-group/survey/pull/20/commits/0adff38bf4c1706dbc27af909cc04560e81567b8)!

Comment (binh-dam-ibigroup) [reply components/QuestionRenderer.tsx]: Move the `<h1>` outside of the `<fieldset>`.

**Review by binh-dam-ibigroup — APPROVED** [review]
One small change then I'm good with this PR.
